### ISSUE TITLE
Use Kubernetes node-local DNS feature to solve DNS intermittent failures in kubernetes

### ISSUE THREAD

**Issue by rata** [kinvolk/lokomotive#240]
Kubernetes node local DNS upstream doc is here [here](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/).

When using kubernetes DNS (`dnsPolicy: ClusterFirst` in the pod.spec, the default) with kube-proxy in iptables mode several problems happen, like: race conditions in the Linux Kernel for connectionless protocols (like UDP that is used for DNS).

**This is killer for latency sensitive apps** as doing a DNS resolution increases a lot the p95 and p99 response time and makes imposible to migrate latency sensitive application to kubernetes. And **you don't need to be doing anything fancy, just a DNS requests** during normal operation (**like to resolve the AWS RDS endpoint) is enough to hit you badly**. I've experienced this myself and is reported upstream in several places, for example [here](https://github.com/kubernetes/kubernetes/issues/56903). My report in the forums ~2 years ago is [here](https://discuss.kubernetes.io/t/kubenet-network-performance-degraded-solved-using-hostnetwork-true-with-unicorn-app/2874).

The website doc is not very clear about this, but it mentions it briefly:

>  Skipping iptables DNAT and connection tracking will help reduce conntrack races and avoid UDP DNS entries filling up conntrack table.

The easiest work-around for this issue *at that time* was not using kubernetes DNSs, if possible, for the app (i.e. it is not using kubrenetes DNS names). This means using `dnsPolicy: Default` in the pod.spec, that avoids entirely the kubernetes DNS (so it doesn't go via kube-proxy that does the iptables rules for the service) and just uses the dns configured in the host `/etc/resolv.conf`file.

This node local DNS feature in kubernetes is supposed to fix this issue, so I think we should see the tradeoff of going for such a feature, and hopefully enable it and solve the problem :)

Regarding the tradeoffs, I'd like to see how much complexity adds to a cluster deployment, if it solves the problem or just reduces the occurence, and what happens if the local cache crashes, etc.

Another thing to take into account might be how kube-proxy in IPVS mode changes this. After all, the issue is about DNAT in connectionless protocols and that might be entirely removed when using kube-proxy in IPVS mode. The node local dns feature might desirable for other reasons, though. Not sure.

**Comment by invidian** [comment]
>The node local dns feature might desirable for other reasons, though. Not sure.

Soo, I spend some time setting node-local-dns over the weekend on my private cluster. I'm running `kube-proxy` in IPVS mode, so I can't say anything about the latency. My main motivation however, was to make DNS resolution functional during single controller node cluster upgrades, as normally CoreDNS runs on controller nodes and ingress controller like nginx requires functional DNS to serve traffic. So if controller node goes down, ingress traffic doesn't work anymore. Having local DNS cache on every node mitigates that, though I'm not sure yet to what degree.

While setting it up, I found that `node-local-dns` docs and manifests seems quite outdated. They still mention `kube-dns` in a lot of place and manifests required a bit of a revamp to get them working properly on my cluster. I ended up creating my own Helm chart to install it.

Other than that, it seems to be working fine.
